vpr/src/noc/noc_traffic_flows.h

```diff
@@ -0,0 +1,307 @@
+/**
+ * @brief Describes a traffic flow within the NoC, which is the communication
+ * between two routers. THe NocTrafficFlows contains a number of this structure
+ * to describe all the communication happening within the NoC.
+ * 
+ */
+struct t_noc_traffic_flow {
+    /** stores the names of the router blocks communicating within this traffic flow*/
+    std::string source_router_module_name;
+    std::string sink_router_module_name;
+
+    /** stores the block id of the two routers blocks communicating within this traffic flow. This can be used to retrieve the block information from the clustered netlist*/
+    ClusterBlockId source_router_cluster_id;
+    ClusterBlockId sink_router_cluster_id;
+
+    /** The bandwidth of the information transferred between the two routers. Units in bps. This parameters will be used to update the link usage in the noc model after routing the traffic flow.*/
+    double traffic_flow_bandwidth;
+
+    /** The maximum allowable time to transmit data between thw two routers. This parameter will be used to evaluate a router traffic flow.*/
+    double max_traffic_flow_latency;
+
+    /** Constructor initializes all variables*/
+    t_noc_traffic_flow(std::string source_router_name, std::string sink_router_name, ClusterBlockId source_router_id, ClusterBlockId sink_router_id, double flow_bandwidth, double max_flow_latency)
+        : source_router_module_name(source_router_name)
+        , sink_router_module_name(sink_router_name)
+        , source_router_cluster_id(source_router_id)
+        , sink_router_cluster_id(sink_router_id)
+        , traffic_flow_bandwidth(flow_bandwidth)
+        , max_traffic_flow_latency(max_flow_latency) {}
+};
+
+class NocTrafficFlows {
+  private:
+    /** contains all the traffic flows provided by the user and their information*/
+    vtr::vector<NocTrafficFlowId, t_noc_traffic_flow> noc_traffic_flows;
+
+    /**
+     * A traffic flow has a source and destination router associated to it. So when either the source or destination router for a given flow is moved, we need to find a new route between them. 
+     *
+     * Therefore, during placement if two router blocks are swapped, then the only traffic flows we need to re-route are the flows where the two routers are either the source or destination routers of those flows. 
+     *
+     * The datastructures below store all traffic flows for each router where the router is a source for the traffic flow. Similarily, there is a another datastructure where all traffic flows are stored for each router where the router is a sink for the traffic flow. The routers are indexed by their ClusterBlockId.
+     *
+     * This is done so that the traffic that need to be re-routed during placement can be quickly found. 
+     */
+    std::unordered_map<ClusterBlockId, std::vector<NocTrafficFlowId>> source_router_associated_traffic_flows;
+    std::unordered_map<ClusterBlockId, std::vector<NocTrafficFlowId>> sink_router_associated_traffic_flows;
```

Comment:
Could group in one unordered_map --> your call.

Reply:
Note but no need to change: can use an unordered_multimap for this case instead of an unordered_map of vectors.

Reply:
Grouped into one datastructure